### PR TITLE
[release/3.1] Port 6.0 fix to AssemblyDependencyResolver to not throw for same file

### PR DESCRIPTION
**Pull request by LakshanF**
Runtime follow up fixes for .NET 3.1 - https://github.com/dotnet/runtime/pull/42055

# Description
There is a problem with resolving multiple files with the same simple name (e.g. Module, module) in case insensitive OS like Windows for AssemblyDependencyResolver. Our internal storage for such cases throws that can cause problems for users who do not expect the error.

# Customer Impact

Reported by customer who was migrating to .Net Core from the Desktop runtime.

After migrating the mixed-mode C++/CLI application from targeting .NET Framework over to .NET Core 3.1, several obstacles have been encountered. One of them is an exception related to adding a duplicate key to a Dictionary in the AssemblyDependencyResolver. Alternatively, they might get an exception about the assembly having already been found with a different extension. Both of these problem point back to .NET Core having a mixture of case sensitivity and insensitivity.

# Regression

No, it has always existed.

# Testing

Added test scenarios. 

A customer who ran into this in 5.0 tried the 6.0 build and reported success.

# Risk

Low.

Although getting casing correct with different OSs are notoriously hard, the resolution was to make AssemblyDependencyResolver consistent with other parts of the runtime by choosing the first simple file name and not throwing afterwards for same simple filename with different casing

The 6.0 issue is https://github.com/dotnet/runtime/issues/37162

### CODE REVIEW

**Review by vitek-karas — COMMENTED** [review]
We should try to port at least one test to validate this.

**Comment by jeffschwMSFT** [comment]
Once you are ready for review, please add the servicing template (here is an example https://github.com/dotnet/runtime/pull/51421)

File: src/corefx/System.Globalization.Native/pal_calendarData.c [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 #include <assert.h>
 #include <stdlib.h>
+#include <stdbool.h>
```
Comment (vitek-karas) [inline src/corefx/System.Globalization.Native/pal_calendarData.c]: This should go in as a separate PR....

File: tests/src/Loader/AssemblyDependencyResolver/AssemblyDependencyResolverTests/AssemblyDependencyResolverTests.cs [hunk 2]
```diff
@@ -130,6 +131,85 @@ public void TestAssembly()
             }
         }
 
+        public void TestAssemblyWithCaseDifferent()
+        {
+            // Testing case sensitive file name resolution in Windows
+            // Host policy returns 2 file paths with the casing changed,
+            // AssemblyDependencyResolver should not throw
+            if(RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (elinor-fung) [inline tests/src/Loader/AssemblyDependencyResolver/AssemblyDependencyResolverTests/AssemblyDependencyResolverTests.cs]: Shouldn't we at least check that we can successfully create the `AssemblyDependencyResolver` and call `ResolveAssemblyToPath` on all platforms instead of just Windows? The behaviour difference due the OS-specific treatment of casing should just be at whether or not `ResolveAssemblyToPath` returns a path, right?

Comment (LakshanF) [reply tests/src/Loader/AssemblyDependencyResolver/AssemblyDependencyResolverTests/AssemblyDependencyResolverTests.cs]: Yes, the source code fix is simply being tolerant on the same simple file name (case different) on the managed side for TPAs. Removed the windows specific check on the test

File: src/corefx/System.Globalization.Native/pal_calendarData.c (hunk 1 above)
Comment (agocke) [reply src/corefx/System.Globalization.Native/pal_calendarData.c]: Sounds good, I was just checking to see if it fixed the mac failures (which it looks like it did). We can remove this commit.

**Comment by agocke** [comment]
@vitek-karas @elinor-fung anything else?

**Review by vitek-karas — APPROVED** [review]

**Comment by agocke** [comment]
Mac fix moved to https://github.com/dotnet/coreclr/pull/28163

**Comment by elinor-fung** [comment]
Can we get a run of the added `AssemblyDependencyResolver` tests on all platforms? They have `CLRTestPriority=1`, so I don't believe they were run with the PR.

**Comment by agocke** [comment]
@elinor-fung Any luck on finding the right thing to run? I agree it looks like only P0 tests are being run, but I can't find a switch to run the rest,

**Comment by elinor-fung** [comment]
Looks like the triggers for it don't include release/3.*. I manually queued a run: https://dev.azure.com/dnceng/public/_build/results?buildId=1108340
I think it should end up reflected in the checks here.

**Comment by elinor-fung** [comment]
Test is failing on Linux: https://helix.dot.net/api/2019-06-17/jobs/552434ac-d9d3-439b-a1d2-2ffe86271a87/workitems/PayloadGroup0/console

```
TestAssemblyWithCaseReversed Start
TestAssemblyWithCaseReversed FAILED:
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> Xunit.Sdk.EqualException: Assert.Equal() Failure
Expected: /home/helixbot/work/A9800949/w/9E9A0883/e/Loader/AssemblyDependencyResolver/AssemblyDependencyResolverTests/AssemblyDependencyResolverTests/TestComponent_15a025f6/TestAssemblyWithCaseReversed.dll
Actual:   (null)
   at Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer`1 comparer) in C:\projects\xunit\src\xunit.assert\Asserts\EqualityAsserts.cs:line 40
   at AssemblyDependencyResolverTests.AssemblyDependencyResolverTests.TestAssemblyWithCaseReversed() in /__w/1/s/tests/src/Loader/AssemblyDependencyResolver/AssemblyDependencyResolverTests/AssemblyDependencyResolverTests.cs:line 195
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at AssemblyDependencyResolverTests.TestBase.<>c__DisplayClass18_1.<RunTestsForInstance>b__1() in /__w/1/s/tests/src/Loader/AssemblyDependencyResolver/AssemblyDependencyResolverTests/TestBase.cs:line 87
   at AssemblyDependencyResolverTests.TestBase.RunSingleTest(Action test, String testName) in /__w/1/s/tests/src/Loader/AssemblyDependencyResolver/AssemblyDependencyResolverTests/TestBase.cs:line 70
```

I expect it is because `AssemblyDependencyResolver.ResolveAssemblyToPath` does a `File.Exists` check on the path, respecting case-sensitivity/insensitivity of the system, so when the first path doesn't match the actual file casing, the assert doesn't hold in case-sensitive scenarios.

**Comment by LakshanF** [comment]
Re-running manually, https://dev.azure.com/dnceng/public/_build/results?buildId=1109754

**Comment by LakshanF** [comment]
The test seems to be passing now

**Review by elinor-fung — APPROVED** [review]

**Comment by agocke** [comment]
@jeffschwMSFT For servicing check

@wtgodbe for guidance on merging. Anything else we should do here?

**Review by jeffschwMSFT — APPROVED** [review]
Approved.  I will take for consideration for 3.1.x.

**Comment by jeffschwMSFT** [comment]
@agocke @LakshanF is there a similar change needed for 5.0.x?

**Comment by wtgodbe** [comment]
If tactics approves this we can merge this once branches are open for June, which will be 5/4 - 5/10 (the Tuesday before Patch Tuesday thru the night before Patch Tuesday). Are the CI failures expected?

**Comment by LakshanF** [comment]
@wtgodbe, the OSX failures that are due to #28163. The test failures in other platforms are not related to this change
